### PR TITLE
Rework training slides

### LINKED ISSUE #357: /unit-testing-add-a-test: distinguish `toEqual` from `toBe`
- make a note of toEqual vs toBe
- reference matching or deep equal

### LINKED ISSUE #358: /unit-testing-getting-tests-to-run: move or remove reference to TestBed

### LINKED ISSUE #366: /unit-testing-spying-on-real-service-providers: Add a reference to debugElement

### LINKED ISSUE #369: /unit-testing-using-async: stick to async
- [ ] Just mention that async function can be used to test asynchronous functions.
- [ ] Change HTTPS to HTTP

### LINKED ISSUE #400: Getting Tests to Run/Cleaning Up Technical Debt
It's too early to have them worry about fixing the module to make all the tests that we don't care about pass.

Let's just ask them to use `fdescribe` on the pipe test to "focus" on it.  We can make note of the fact that if you start to build an app without paying attention to generated unit tests, you'll definitely have a bunch of failing tests.

### PR DESCRIPTION
**Pull request by mdegani**
Remove extra "pipe" (#355)
Rework testing slides (#363)

fixes #366 (we explain DebugElement)
fixes #358 (we explain reason for using TestBed and show a simple example)
fixes #357 by explaining the difference between `toBe` and `toEqual`
fixes #369 by indicating the async is used to test asynchronous functions
fixes #400 by not refactoring, just using `fdescribe`

makes #371 no longer relevant

### CODE REVIEW

File: slides/src/content/unit-testing.md [hunk 1]
```diff
@@ -57,7 +68,7 @@ Note the following in the generated tests:
 <!-- .slide: id="unit-testing-testing-a-pipe" -->
 ## Testing a Pipe
 
-- `ng generate pipe pipe capitalize`
+- `ng generate pipe capitalize`
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: good catch

File: slides/src/content/unit-testing.md [hunk 2]
```diff
@@ -89,17 +100,17 @@ describe('CapitalizePipe', () => {
...
-- We will need a fixture for each test, so don't duplicate that code
+- We will need an instance for each test, so instantiate it in the `beforeEach` to avoid duplicating code
 
 #####_src/app/capitalize.pipe.spec.ts_
 ```ts
 describe('CapitalizePipe', () => {
 
   let pipe;
-
+
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: please strip trailing whitespace off lines (it sometimes makes Markdown think we want a `<br>`

File: slides/src/content/unit-testing.md [hunk 3]
```diff
@@ -115,9 +126,11 @@ describe('CapitalizePipe', () => {
 <!-- .slide: id="unit-testing-add-a-test" -->
 ## Add a Test
 
+- Use `fdescribe` to tell Jasmine to "**f**ocus" on this test -- we have several broken tests that we'll ignore for now.
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: 1.  Maybe turn the second half of the sentence (after `--`) into a sub-point?
2.  Do I have to edit the test file whenever I want to focus on particular tests, or can I specify tests from the command line when I run things?

File: slides/src/content/unit-testing.md [hunk 4]
```diff
@@ -128,117 +141,87 @@ describe('CapitalizePipe', () => {
...
 ---
-<!-- .slide: id="unit-testing-getting-tests-to-run" -->
-## Getting Tests to Run
+<!-- .slide: id="unit-testing-instantiating-component--or-service" -->
+## Instantiating the Component or Service
 
-- Run `ng test`: 4 failures
-- Have to include our dependencies in `app.component.spec.ts`
-  - And in `to-do-list.component.spec.ts`
+- We currently do not display the number of todos in `AppComponent`.
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: good example

File: slides/src/content/unit-testing.md [hunk 5]
```diff
@@ -128,117 +141,87 @@ describe('CapitalizePipe', () => {
...
 
-#####_src/app/to-do.service.ts_
-```ts
-export class ToDoService {
-  // ...as before...
-  itemCount() {
-    return this.items.length;
-  }
-}
+});
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: Dunno why Git is interleaving the diff this way, but I *think* this code fragment is good.

File: slides/src/content/unit-testing.md [hunk 6]
```diff
@@ -128,117 +141,87 @@ describe('CapitalizePipe', () => {
...
 
 ---
 <!-- .slide: id="unit-testing-mocking-dependencies" -->
 ## Mocking Dependencies
 
 - We need an object that matches the "shape" of `TodoService` to test `AppComponent`
-- The mock object must have `onNewItem` and `itemCount` methods
-- `onNewItem` doesn't actually need to do anything
-- `itemCount` can return a fixed value
+- Note the use of `as` to assert the type of our mock.
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: Are we asserting the type or doing a typecast?

File: slides/src/content/unit-testing.md [hunk 7]
```diff
@@ -128,117 +141,87 @@ describe('CapitalizePipe', () => {
...
 ```
 
-<!-- preview: http://plnkr.co/edit/XUM8Gfz08nfbQf1BhDN1?p=preview FIXME update -->
-
 ---
 <!-- .slide: id="unit-testing-when-to-use-testbed" -->
 ## When to Use TestBed
 
-`TestBed` is helpful when:
+`TestBed` is a class that creates a real Angular runtime for the purposes of testing Angular elements. It is helpful when:
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: start a new bullet point with "It is helpful"

File: slides/src/content/unit-testing.md [hunk 8]
```diff
@@ -248,70 +231,30 @@ mockTodoService = {
...
-- ...but sometimes Angular is not able to resolve all of the required dependencies
-- We have more control when we mock things...
-- ...but if mocks get too complicated, they can be expensive to maintain
 
----
 <!-- .slide: id="unit-testing-testing-components-using-componentfixture" -->
 ## Testing Components Using ComponentFixture
 
-Here is how we obtain the instance from the fixture:
+- `ComponentFixture` is an testing utility that gives us access and control over the things we are testing.
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: "an testing" => "a testing"

File: slides/src/content/unit-testing.md [hunk 9]
```diff
@@ -248,70 +231,30 @@ mockTodoService = {
...
+- Here is how we obtain the instance of our declared component from the fixture.
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: Rather than "Here is how we...", prefer "We obtain X by..." and be explicit - "by referencing in providers" ? or...?

File: slides/src/content/unit-testing.md [hunk 10]
```diff
@@ -332,39 +275,63 @@ beforeEach(() => {
...
 ---
-<!-- .slide: id="unit-testing-change-detection-1" -->
-## Change Detection
 
-- We must tell Angular when to run change detection during our tests
-- Use the fixture to do this:
+<!-- .slide: id="unit-testing-shallow-rendering" -->
+## Shallow Rendering
+
+- The above test fails because `AppComponent` itself uses components which we have not declared.
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: remove "itself"

File: slides/src/content/unit-testing.md [hunk 11]
```diff
@@ -332,39 +275,63 @@ beforeEach(() => {
...
+- We could provide mocks for these components, or we can do a shallow render by adding the `CUSTOM-ELEMENTS-SCHEMA`
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: 1. "or we can do" => "or do"
2. is `CUSTOM-ELEMENTS-SCHEMA` hyphenated or underscored? looks in the code below like underscores

File: slides/src/content/unit-testing.md [hunk 12]
```diff
@@ -332,39 +275,63 @@ beforeEach(() => {
...
-fixture = TestBed.createComponent(AppComponent);
-fixture.detectChanges()
+TestBed.configureTestingModule({
+  schemas: [ CUSTOM_ELEMENTS_SCHEMA ]
+})
 ```
 
+Notes:
+
+- `NO_ERRORS_SCHEMA` is also commonly used for shallow rendering.
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: 1.  Remove "Notes:"
2.  We haven't introduced schemas previously - add a slide explaining what they are?

File: slides/src/content/unit-testing.md [hunk 13]
```diff
@@ -332,39 +275,63 @@ beforeEach(() => {
...
-    { provide: ComponentFixtureAutoDetect, useValue: true }
-  ]
+  declarations: [ AppComponent ],
+  schemas: [ CUSTOM_ELEMENTS_SCHEMA ],
+  providers:[ ToDoService ]
+});
+
+const toDoService = TestBed.get(ToDoService);
+toDoService.items = ['item 1', 'item 2', 'item 3'];
+
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: remove blank line

File: slides/src/content/unit-testing.md [hunk 14]
```diff
@@ -332,39 +275,63 @@ beforeEach(() => {
...
+```
+
+---
+
+## Importing the Real Module
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: slide ID?

File: slides/src/content/unit-testing.md [hunk 15]
```diff
@@ -332,39 +275,63 @@ beforeEach(() => {
...
 });
 ```
 
 ---
 <!-- .slide: id="unit-testing-querying-native-elements" -->
-## Querying Native Elements
+## Integration Testing by Querying Native Elements
 
-- Once change detection has run, we can check that the expeced output was rendered to the DOM
+- Importing the AppModule, override the `ToDoService` by providing the mock as before.
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: back ticks around AppModule

File: slides/src/content/unit-testing.md [hunk 16]
```diff
@@ -378,20 +345,51 @@ describe('AppComponent', () => {
...
+});
+```
+
+---
+
 <!-- .slide: id="unit-testing-dealing-with-async" -->
 ## Dealing With Asynchronous Behavior in Tests
 
-- Angular CLI uses the asynchronous function `compileComponents` and wraps it in the `async` function
+- Angular tests generated by the CLI use the asynchronous function `compileComponents` and wraps it in the `async` function
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: "wrap" instead of "wraps"

File: slides/src/content/unit-testing.md [hunk 17]
```diff
@@ -498,74 +499,62 @@ export class TodoService {
...
-```
+<!-- .slide: id="unit-testing-light-http-mock" -->
+## Providing a Light HTTP Mock
 
----
-<!-- .slide: id="unit-testing-creating-and-inject-mock-object-2" -->
-## Creating and Injecting the Mock Object
-
-- We can create a light mock of the `Http` service:
+- A light mock can be used in place of the actual Http service
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: 1.  back ticks around Http
2.  What's a "light mock"?  Light as in "lightweight"?

File: slides/src/content/unit-testing.md [hunk 18]
```diff
@@ -498,74 +499,62 @@ export class TodoService {
...
+- Note the use of `as` to assert the type of our mock.
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: asserting type or doing typecast? or are they the same in TypeScript?

File: slides/src/content/unit-testing.md [hunk 19]
```diff
@@ -498,74 +499,62 @@ export class TodoService {
...
+- A jasmine spy provides the mock implementation of the get method.
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: Jasmine (capital J)

File: slides/src/content/unit-testing.md [hunk 20]
```diff
@@ -498,74 +499,62 @@ export class TodoService {
...
-beforeEach(() => {
-  mockHttp = { get: () => { } };
+describe('toDoService', () => {
+  let toDoService: ToDoService;
+  let mockHttp: Http;
 
-});
-```
+  beforeEach(() => {
+
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: remove blank lines (or does that make the code too ugly)?

File: slides/src/content/unit-testing.md [hunk 21]
```diff
@@ -498,74 +499,62 @@ export class TodoService {
...
 ---
-<!-- .slide: id="unit-testing-asserting-on-the-request-and-response" -->
-## Asserting on the Request and Response
+<!-- .slide: id="unit-testing-spying-to-check-url" -->
+## Using the Spy to Check the URL
 
-- This method still allows us to check that the service
-  - Has requested the right URL
-  - Returns that expected data
+- The following test checks that the correct url is being used.
```
Comment (gvwilson) [inline slides/src/content/unit-testing.md]: URL instead of url

File: slides/src/content/unit-testing.md (hunk 3 above)
Comment (mdegani) [inline slides/src/content/unit-testing.md]: It's one thought, so I've reworded it.

I don't know if Karma/Jasmine has a filter feature that let's you select tests while in watch most like jest does.  Added issue #498 for this.

File: slides/src/content/unit-testing.md (hunk 6 above)
Comment (mdegani) [inline slides/src/content/unit-testing.md]: This is just an assertion for static type checking and has no impact on runtime code. Rather than changing the type, it's telling TypeScript to *act* like it's the other type.

File: slides/src/content/unit-testing.md (hunk 11 above)
Comment (mdegani) [inline slides/src/content/unit-testing.md]: Yes, should be snake case.  Making that fix.

File: slides/src/content/unit-testing.md (hunk 12 above)
Comment (mdegani) [inline slides/src/content/unit-testing.md]: Schemas are just a way of telling angular not complain about unknown elements. Is it worth a slide?

File: slides/src/content/unit-testing.md (hunk 17 above)
Comment (mdegani) [inline slides/src/content/unit-testing.md]: I got "light mock" from the gitbook and I thought it was an appropriate way to describe it. "Lightweight" to me seems to be counter to something that is heavy and cumbersome. Here is used as in the phrase "this slide if light on detail". Should we make this a discussion?

File: slides/src/content/unit-testing.md (hunk 18 above)
Comment (mdegani) [inline slides/src/content/unit-testing.md]: Just an assertion, not casting.

File: slides/src/content/unit-testing.md (hunk 20 above)
Comment (mdegani) [inline slides/src/content/unit-testing.md]: I agree, remove the lines.

**Comment by mdegani** [comment]
@gvwilson I think I addressed everything. Sure I'll go through the slides on Monday, see how the work.

**Comment by gvwilson** [comment]
Probably not worth a slide, but maybe a one-liner? My first guess was it
was something to do with databases, which confused me.
